src/analysis/retail/warrior/shared/modules/debuggers/RageCountDebugger.tsx

```diff
@@ -0,0 +1,137 @@
+import getRage from 'analysis/retail/warrior/shared/utils/getRage';
+import { formatDuration } from 'common/format';
+import { BadColor } from 'interface/guide';
+import Analyzer, { Options } from 'parser/core/Analyzer';
+import Events, { AbilityEvent, AnyEvent, EventType, ResourceChangeEvent } from 'parser/core/Events';
+
+const LAST_EVENTS_TO_KEEP = 12;
+
+/**
+ * Set to tru to print out all events and the rage changes that occur. Can be
+ * helpful to look into when rage mismatches. If false, only prints Rage mismatches
+ */
+const PRINT_ALL = false;
+
+/**
+ * Debugger that keeps track of rage and prints an error whenever rage between two logged
+ * events does not match any resource changes that have occurred.
+ *
+ * If we recieve a logged event reporting that rage is at 80, and then later recieve an event
+ * that reports that rage is at 90, but we have not seen any resourceChange events that would
+ * account for this, we will print an error.
+ */
+class RageCountDebugger extends Analyzer {
+  constructor(options: Options) {
+    super(options);
+
+    // Kepe track of rage
+    let expectedRage: number | undefined = undefined;
+
+    const lastEvents: AnyEvent[] = [];
+
+    this.addEventListener(Events.any, (event: AnyEvent) => {
+      if (event.type === EventType.SpendResource) {
+        // This is a fabricated event from the base ResourceTracker, which is not useful for us
+        return;
+      }
+
+      lastEvents.push(event);
+      if (lastEvents.length > LAST_EVENTS_TO_KEEP) {
+        lastEvents.shift();
+      }
+
+      const rage = getRage(event, this.selectedCombatant);
+
+      let warning = false;
+      const parts: string[] = [];
+      if ('resourceChange' in event) {
+        parts.push(`Resource Change: ${event.resourceChange}`);
+        if ('waste' in event && event.waste !== 0) {
+          parts.push(`Waste: ${event.waste}`);
+        }
+        if (expectedRage != null) {
+          expectedRage += event.resourceChange - (event as any).waste || 0;
+        }
+      }
+
+      if (expectedRage != null && rage?.amount != null) {
+        if (rage.amount === expectedRage) {
+          // This doesn't really work in the debug UI.
+          // this.addDebugAnnotation(event, {
+          //   summary: `Rage correct: ${rage.amount}`,
+          //   color: 'green',
+          // });
```

Comment:
are there just too many events when you enable this?

Reply:
Yeah it's for every single event that logs the current rage, which is all spenders and gains. The debug UI renders one row of dots per minutes, so if you get 100s of them, they squeeze and overflow. Not sure if the debugui should handle that many dots either. I've only just now discovered the debug ui